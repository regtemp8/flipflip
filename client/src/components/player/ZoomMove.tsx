import React, { useRef, type PropsWithChildren, useMemo } from 'react'
import { animated, useSpring } from '@react-spring/web'
import { ZoomMoveData } from '../../store/player/ContentPreloadService'
import { getEaseFunction } from '../../data/utils'

interface EffectProps {
  data: ZoomMoveData
  show: boolean
  isPlaying: boolean
  className?: string
}

function Effect(props: PropsWithChildren<EffectProps>) {
  const { data, show, isPlaying, className, children } = props

  const _prevShow = useRef<boolean>(false)

  const animation = useMemo(() => {
    const toTransform: string[] = []
    const fromTransform: string[] = []
    if (data.translateX !== 0 && data.translateY !== 0) {
      fromTransform.push('translate(0%, 0%)')
      toTransform.push(`translate(${data.translateX}%, ${data.translateY}%)`)
    } else if (data.translateX !== 0) {
      fromTransform.push('translateX(0%)')
      toTransform.push(`translateX(${data.translateX}%)`)
    } else if (data.translateY !== 0) {
      fromTransform.push('translateY(0%)')
      toTransform.push(`translateY(${data.translateY}%)`)
    }
    if (data.scaleFrom !== 1 || data.scaleTo !== 1) {
      fromTransform.push(`scale(${data.scaleFrom})`)
      toTransform.push(`scale(${data.scaleTo})`)
    }

    let easing: ((time: number) => number) | undefined = undefined
    if (data.easing != null) {
      const { ea, exp, amp, per, ov } = data.easing
      easing = getEaseFunction(ea, exp, amp, per, ov)
    }

    return {
      from: {
        transform: fromTransform.join(' ')
      },
      to: {
        transform: toTransform.join(' ')
      },
      config: {
        duration: data.duration,
        easing
      }
    }
  }, [data])

  const reset = _prevShow.current !== show && show
  _prevShow.current = show
  const style = useSpring({
    ...animation,
    reset,
    pause: !isPlaying
  })
  return (
    <animated.div className={className} style={style}>
      {children}
    </animated.div>
  )
}

export interface ZoomMoveProps {
  show: boolean
  isPlaying: boolean
  data?: ZoomMoveData
  className?: string
}

export default function ZoomMove(props: PropsWithChildren<ZoomMoveProps>) {
  const { show, isPlaying, data, className, children } = props
  if (data != null) {
    return (
      <Effect
        data={data}
        show={show}
        isPlaying={isPlaying}
        className={className}
      >
        {children}
      </Effect>
    )
  } else {
    return <>{children}</>
  }
}

;(ZoomMove as any).displayName = 'ZoomMove'
