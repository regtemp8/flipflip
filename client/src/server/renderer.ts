import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
