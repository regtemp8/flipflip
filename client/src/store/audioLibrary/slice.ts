import { createSlice, type PayloadAction } from '@reduxjs/toolkit'

interface AudioLibraryState {
  error: boolean
  loadingMetadata: boolean
  loadingSources: boolean
  drawerOpen: boolean
  selectedTags: string[]
  openMenu?: string
  playlistID?: number
  importURL?: string
}

export const initialAudioLibraryState: AudioLibraryState = {
  error: false,
  loadingMetadata: false,
  loadingSources: false,
  drawerOpen: false,
  selectedTags: []
}

export const audioLibrarySlice = createSlice({
  name: 'audioLibrary',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialAudioLibraryState,
  reducers: {
    setError: (state, action: PayloadAction<boolean>) => {
      state.error = action.payload
    },
    setLoadingMetadata: (state, action: PayloadAction<boolean>) => {
      state.loadingMetadata = action.payload
    },
    setLoadingSources: (state, action: PayloadAction<boolean>) => {
      state.loadingSources = action.payload
    },
    setOpenMenu: (state, action: PayloadAction<string | undefined>) => {
      state.openMenu = action.payload
    },
    setPlaylistID: (state, action: PayloadAction<number | undefined>) => {
      state.playlistID = action.payload
    },
    setDrawerOpen: (state, action: PayloadAction<boolean>) => {
      state.drawerOpen = action.payload
    },
    setImportURL: (state, action: PayloadAction<string | undefined>) => {
      state.importURL = action.payload
    },
    setSelectedTags: (state, action: PayloadAction<string[]>) => {
      state.selectedTags = action.payload
    },
    setAudioLibraryPlaylistID: (state, action: PayloadAction<string>) => {
      state.playlistID = Number(action.payload)
    }
  }
})

export const {
  setError,
  setLoadingMetadata,
  setLoadingSources,
  setOpenMenu,
  setPlaylistID,
  setDrawerOpen,
  setImportURL,
  setSelectedTags,
  setAudioLibraryPlaylistID
} = audioLibrarySlice.actions

export default audioLibrarySlice.reducer
