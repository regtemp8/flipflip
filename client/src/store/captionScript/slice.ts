import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type CaptionScript from './CaptionScript'
import {
  type EntryState,
  type EntryUpdate,
  setEntrySlice,
  setEntry,
  updateEntry
} from '../EntryState'
import { type FontSettingsType } from './FontSettings'
import { getCaptionScript, getCaptionScriptFontSettings } from './selectors'

export interface FontSettingsUpdate<T> extends EntryUpdate<T> {
  type: FontSettingsType
}

export const initialCaptionScriptState: EntryState<CaptionScript> = {
  name: 'captionScriptSlice',
  nextID: 1,
  entries: {}
}
export const captionScriptSlice = createSlice({
  name: 'captionScripts',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialCaptionScriptState,
  reducers: {
    setCaptionScriptSlice: (
      state,
      action: PayloadAction<EntryState<CaptionScript>>
    ) => {
      setEntrySlice(state, action.payload)
    },
    setCaptionScript: (state, action: PayloadAction<CaptionScript>) => {
      setEntry(state, action.payload)
    },
    updateCaptionScript: (state, action: PayloadAction<CaptionScript>) => {
      updateEntry(state, action.payload)
    },
    setCaptionScriptScript: (
      state,
      action: PayloadAction<EntryUpdate<string>>
    ) => {
      getCaptionScript(state, action.payload.id).script = action.payload.value
    },
    setCaptionScriptURL: (
      state,
      action: PayloadAction<EntryUpdate<string>>
    ) => {
      getCaptionScript(state, action.payload.id).url = action.payload.value
    },
    setCaptionScriptStopAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).stopAtEnd =
        action.payload.value
    },
    setCaptionScriptNextSceneAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).nextSceneAtEnd =
        action.payload.value
    },
    setCaptionScriptSyncWithAudio: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).syncWithAudio =
        action.payload.value
    },
    setCaptionScriptOpacity: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getCaptionScript(state, action.payload.id).opacity = action.payload.value
    },
    setCaptionScriptFontSettingsBorder: (
      state,
      action: PayloadAction<FontSettingsUpdate<boolean>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).border = action.payload.value
    },
    setCaptionScriptFontSettingsColor: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).color = action.payload.value
    },
    setCaptionScriptFontSettingsBorderColor: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).borderColor = action.payload.value
    },
    setCaptionScriptFontSettingsFontFamily: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).fontFamily = action.payload.value
    },
    setCaptionScriptFontSettingsFontSize: (
      state,
      action: PayloadAction<FontSettingsUpdate<number>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).fontSize = action.payload.value
    },
    setCaptionScriptFontSettingsBorderPx: (
      state,
      action: PayloadAction<FontSettingsUpdate<number>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).borderpx = action.payload.value
    },
    setCaptionScriptTags: (
      state,
      action: PayloadAction<EntryUpdate<number[]>>
    ) => {
      state.entries[action.payload.id].tags = action.payload.value
    },
    setCaptionScriptToggleTag: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      const script = state.entries[action.payload.id]
      const index = script.tags.indexOf(action.payload.value)
      if (index === -1) {
        script.tags.push(action.payload.value)
      } else {
        script.tags.splice(index, 1)
      }
    },
    setCaptionScriptMarked: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).marked = action.payload.value
    }
  }
})

export const {
  setCaptionScriptSlice,
  setCaptionScript,
  updateCaptionScript,
  setCaptionScriptScript,
  setCaptionScriptURL,
  setCaptionScriptStopAtEnd,
  setCaptionScriptNextSceneAtEnd,
  setCaptionScriptSyncWithAudio,
  setCaptionScriptOpacity,
  setCaptionScriptFontSettingsBorder,
  setCaptionScriptFontSettingsColor,
  setCaptionScriptFontSettingsBorderColor,
  setCaptionScriptFontSettingsFontFamily,
  setCaptionScriptFontSettingsFontSize,
  setCaptionScriptFontSettingsBorderPx,
  setCaptionScriptTags,
  setCaptionScriptToggleTag,
  setCaptionScriptMarked
} = captionScriptSlice.actions

export default captionScriptSlice.reducer
