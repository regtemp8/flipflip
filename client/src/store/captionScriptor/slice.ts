import { createSlice, type PayloadAction } from '@reduxjs/toolkit'

interface CaptionScriptorState {
  sceneID: number
  captionScriptID: number
  sceneScripts: string[]
}

export const initialCaptionScriptorState: CaptionScriptorState = {
  sceneID: 0,
  captionScriptID: -1,
  sceneScripts: []
}
export const captionScriptorSlice = createSlice({
  name: 'captionScriptor',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialCaptionScriptorState,
  reducers: {
    setCaptionScriptorSceneID: (state, action: PayloadAction<number>) => {
      state.sceneID = action.payload
    },
    setCaptionScriptorSceneScripts: (
      state,
      action: PayloadAction<string[]>
    ) => {
      state.sceneScripts = action.payload
    },
    setCaptionScriptorCaptionScriptID: (
      state,
      action: PayloadAction<number | undefined>
    ) => {
      state.captionScriptID = action.payload ?? -1
    }
  }
})

export const {
  setCaptionScriptorSceneID,
  setCaptionScriptorSceneScripts,
  setCaptionScriptorCaptionScriptID
} = captionScriptorSlice.actions

export default captionScriptorSlice.reducer
