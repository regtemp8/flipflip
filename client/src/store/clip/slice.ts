import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type Clip from './Clip'
import {
  type EntryState,
  type EntryUpdate,
  setEntrySlice,
  setEntry
} from '../EntryState'

export const initialClipState: EntryState<Clip> = {
  name: 'clipSlice',
  nextID: 1,
  entries: {}
}

export default function createClipReducer(clipState?: EntryState<Clip>) {
  return createClipSlice(clipState).reducer
}

function createClipSlice(clipState?: EntryState<Clip>) {
  const initialState = clipState ?? initialClipState
  return createSlice({
    name: 'clips',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
      setClipSlice: (state, action: PayloadAction<EntryState<Clip>>) => {
        setEntrySlice(state, action.payload)
      },
      setClip: (state, action: PayloadAction<Clip>) => {
        setEntry(state, action.payload)
      },
      setClips: (state, action: PayloadAction<Clip[]>) => {
        action.payload.forEach((s) => setEntry(state, s))
      },
      setClipToggleTag: (state, action: PayloadAction<EntryUpdate<number>>) => {
        const clip = state.entries[action.payload.id]
        const index = clip.tags.indexOf(action.payload.value)
        if (index === -1) {
          clip.tags.push(action.payload.value)
        } else {
          clip.tags.splice(index, 1)
        }
      },
      setClipVolume: (
        state,
        action: PayloadAction<EntryUpdate<number | undefined>>
      ) => {
        state.entries[action.payload.id].volume = action.payload.value
      },
      setClipTags: (state, action: PayloadAction<EntryUpdate<number[]>>) => {
        state.entries[action.payload.id].tags = action.payload.value
      },
      setClipStartEnd: (
        state,
        action: PayloadAction<EntryUpdate<number[]>>
      ) => {
        const clip = state.entries[action.payload.id]
        clip.start = action.payload.value[0]
        clip.end = action.payload.value[1]
      }
    }
  })
}

export const {
  setClipSlice,
  setClip,
  setClips,
  setClipToggleTag,
  setClipTags,
  setClipVolume,
  setClipStartEnd
} = createClipSlice().actions
