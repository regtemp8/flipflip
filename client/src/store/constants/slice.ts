import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { initialSystemConstants } from './SystemConstants'
import { SystemConstants, copy } from 'flipflip-common'

export default function createConstantsReducer(constants?: SystemConstants) {
  return createConstantsSlice(constants).reducer
}

function createConstantsSlice(constants?: SystemConstants) {
  let initialState = copy<SystemConstants>(initialSystemConstants)
  if (constants != null) {
    initialState = Object.assign(initialState, constants)
  }

  return createSlice({
    name: 'constants',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
      setConstants: (state, action: PayloadAction<SystemConstants>) => {
        return action.payload
      }
    }
  })
}

export const { setConstants } = createConstantsSlice().actions
