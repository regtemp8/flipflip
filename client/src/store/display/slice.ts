import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import type Display from './Display'
import {
  setEntry,
  type EntryState,
  EntryUpdate,
  getEntry,
  deleteEntry
} from '../EntryState'

export interface DisplayViewUpdate<T> extends EntryUpdate<T> {
  index: number
}

export const initialDisplayState: EntryState<Display> = {
  name: 'displaySlice',
  nextID: 1,
  entries: {}
}

export default function createDisplayReducer(
  displayState?: EntryState<Display>
) {
  return createDisplaySlice(displayState).reducer
}

function createDisplaySlice(displayState?: EntryState<Display>) {
  const initialState = displayState ?? initialDisplayState
  return createSlice({
    name: 'displays',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
      setDisplaySlice: (state, action: PayloadAction<EntryState<Display>>) => {
        return action.payload
      },
      setDisplay: (state, action: PayloadAction<Display>) => {
        setEntry(state, action.payload)
      },
      deleteDisplay: (state, action: PayloadAction<number>) => {
        deleteEntry(state, action.payload)
      },
      setDisplayName: (state, action: PayloadAction<EntryUpdate<string>>) => {
        const { id, value } = action.payload
        getEntry(state, id).name = value
      },
      setDisplayAddView: (
        state,
        action: PayloadAction<EntryUpdate<number>>
      ) => {
        const { id, value } = action.payload
        const entry = getEntry(state, id)
        entry.views.unshift(value)
        entry.selectedView = value
      },
      setDisplayRemoveView: (
        state,
        action: PayloadAction<EntryUpdate<number>>
      ) => {
        const { id, value } = action.payload
        const entry = getEntry(state, id)
        const index = entry.views.indexOf(value)
        entry.views.splice(index, 1)
        entry.selectedView =
          entry.views[Math.min(index, entry.views.length - 1)]
      },
      setDisplayViewsListYOffset: (
        state,
        action: PayloadAction<EntryUpdate<number>>
      ) => {
        const { id, value } = action.payload
        getEntry(state, id).displayViewsListYOffset = value
      },
      setDisplaySelectedView: (
        state,
        action: PayloadAction<EntryUpdate<number | undefined>>
      ) => {
        const { id, value } = action.payload
        getEntry(state, id).selectedView = value
      },
      swapDisplayViews: (
        state,
        action: PayloadAction<EntryUpdate<number[]>>
      ) => {
        const { id, value } = action.payload
        const entry = getEntry(state, id)
        const oldID = entry.views[value[0]]
        entry.views[value[0]] = entry.views[value[1]]
        entry.views[value[1]] = oldID
      }
    }
  })
}

export const {
  setDisplaySlice,
  setDisplay,
  deleteDisplay,
  setDisplayName,
  setDisplayAddView,
  setDisplayRemoveView,
  setDisplayViewsListYOffset,
  setDisplaySelectedView,
  swapDisplayViews
} = createDisplaySlice().actions
