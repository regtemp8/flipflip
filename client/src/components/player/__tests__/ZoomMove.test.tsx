import React from 'react'
import { describe, it, expect } from '@jest/globals'
import renderer from 'react-test-renderer'
import ZoomMove from '../ZoomMove'
import TestProvider from '../../../util/TestProvider'
import store from '../../../store/store'
import { setScene } from '../../../store/scene/slice'
import { newScene } from '../../../store/scene/Scene'

describe('ZoomMove', () => {
  it('should match snapshot', () => {
    const sceneID = 3
    store.dispatch(setScene(newScene({ id: sceneID })))

    const component = renderer.create(
      <TestProvider store={store}>
        <ZoomMove
          sceneID={sceneID}
          reset={false}
          timeToNextFrame={0}
          currentAudio={null}
        >
          <p>Test</p>
        </ZoomMove>
      </TestProvider>
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
