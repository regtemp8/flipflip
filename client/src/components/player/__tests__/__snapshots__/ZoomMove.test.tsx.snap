// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoomMove should match snapshot 1`] = `
<div
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": "translate(0%, 0%) scale(1)",
      "zIndex": 2,
    }
  }
>
  <p>
    Test
  </p>
</div>
`;
